compiler/bindings/python/iree/build/args.py

```diff
@@ -0,0 +1,187 @@
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import Callable, Generator, TypeVar
+
+import argparse
+import contextlib
+import functools
+import inspect
+import threading
+
+from typing import Callable
+
+_locals = threading.local()
+_ALL_ARG_REGISTRARS: list[Callable[[argparse.ArgumentParser], None]] = []
+_ALL_ARG_HANDLERS: list[Callable[[argparse.Namespace], None]] = []
+
+
+def register_arg_parser_callback(registrar: Callable[[argparse.ArgumentParser], None]):
+    """Decorator that adds a global argument registration callback.
+
+    This callback will be invoked when a new ArgumentParser is constructed.
+    """
+    _ALL_ARG_REGISTRARS.append(registrar)
+    return registrar
+
+
+def register_arg_handler_callback(handler: Callable[[argparse.Namespace], None]):
+    """Decorator that registers a handler to be run on global arguments at startup."""
+    _ALL_ARG_HANDLERS.append(handler)
+    return handler
+
+
+def configure_arg_parser(p: argparse.ArgumentParser):
+    """Invokes all callbacks from `register_arg_parser_callback` on the parser."""
+    for callback in _ALL_ARG_REGISTRARS:
+        callback(p)
+
+
+def run_global_arg_handlers(ns: argparse.Namespace):
+    """Invokes all global argument handlers."""
+    for h in _ALL_ARG_HANDLERS:
+        h(ns)
+
+
+@contextlib.contextmanager
+def argument_namespace_context(ns: argparse.Namespace):
+    """Establish that given namespace as the current namespace for this thread.
+
+    Note that as a thread local, this does not propagate to child threads or
+    sub-processes. This means that all argument management must be done during
+    action setup and action invocations will not typically have access to args.
+    """
```

Comment:
Thanks for the docstrings! I filed https://github.com/iree-org/iree/issues/19019 to make sure this finds its way to https://iree-python-api.readthedocs.io/en/latest/index.html. Plenty of work needed to get that docs site up to date and useful...